### PR TITLE
Implement a tree view of the objects in the scene editor

### PR DESCRIPTION
**Pull request by AlexandreSi**
This PR implements object folders as a way to organize the objects of a scene or an external layout.

It also removes the object tags feature as its success as an organization feature was not proven.

It introduces a new TreeView UI component, not relying on Material UI for performance reasons.
This new component:
- is virtualized
- makes it easy to drag&drop elements on desktop and on mobile
- has a notion of hierarchy (folders)

Other changes:
- When creating/editing an action/condition, if a folder or an object groups matches the search the content of those is shown 
- Some shortcuts have been added to the objects list to navigate with the arrow keys and to be able to delete an item with Backspace

[Story for the TreeView component](https://gdevelop-storybook.s3.amazonaws.com/object-folders/latest/index.html?path=/story/ui-building-blocks-treeview--default)

### CODE REVIEW

File: newIDE/app/package.json [hunk 1]
```diff
@@ -85,6 +85,7 @@
     "react-sortable-tree": "2.6.2",
     "react-test-renderer": "16.14.0",
     "react-virtualized": "9.21.1",
+    "react-window": "^1.8.9",
```
Comment (4ian) [inline newIDE/app/package.json:88]: Probably a good idea to use a fixed version

File: newIDE/app/src/UI/DragAndDrop/DragAndDropContextProvider.js [hunk 2]
```diff
@@ -1,8 +1,22 @@
 import { Component } from 'react';
-import MultiBackend from 'react-dnd-multi-backend';
-import HTML5toTouch from 'react-dnd-multi-backend/lib/HTML5toTouch';
+import HTML5Backend from 'react-dnd-html5-backend';
+import TouchBackend from 'react-dnd-touch-backend';
+import MultiBackend, { TouchTransition } from 'react-dnd-multi-backend';
 import { DragDropContext } from 'react-dnd';
 
+const HTML5toTouch = {
```
Comment (4ian) [inline newIDE/app/src/UI/DragAndDrop/DragAndDropContextProvider.js:7]: Interesting, why this change? Should be tested in the mobile app, which had issue with the transition from what I remember

Comment (AlexandreSi) [reply newIDE/app/src/UI/DragAndDrop/DragAndDropContextProvider.js:7]: It's to be able to define `delayTouchStart` parameter that defaults to 0.

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        childObjectFolderOrObject->UnserializeFrom(
+            project, childrenElements.GetChild(i), objectsContainer);
+        childObjectFolderOrObject->parent = this;
+        children.push_back(std::move(childObjectFolderOrObject));
+      }
+    }
+  } else {
+    folderName = "";
+    object =
+        &objectsContainer.GetObject(element.GetStringAttribute("objectName"));
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:232]: What happens if the object does not exist? I would protect against it by showing a `gd::LogWarning` or `gd::LogError`

Comment (4ian) [reply Core/GDCore/Project/ObjectFolderOrObject.cpp:232]: and keep a nullptr to show that "there is an issue here, hope the rest of the code will handle this properly".

File: Core/GDCore/Project/Project.cpp [hunk 4]
```diff
@@ -1000,6 +1005,7 @@ void Project::SerializeTo(SerializerElement& element) const {
 
   resourcesManager.SerializeTo(element.AddChild("resources"));
   SerializeObjectsTo(element.AddChild("objects"));
+  SerializeFoldersTo(element.AddChild("folderStructure"));
```
Comment (4ian) [inline Core/GDCore/Project/Project.cpp:1008]: I think that should be `objectsFolderStructure`

**Review by 4ian — COMMENTED** [review]
Just some quick notes, I have to review more later.

File: newIDE/app/src/ObjectGroupsList/index.js [hunk 5]
```diff
@@ -67,430 +80,521 @@ type Props = {|
...
+          return;
+        }
+
+        if (beforeSetAsGlobalGroup && !beforeSetAsGlobalGroup(groupName)) {
+          return;
+        }
+
+        const answer = await showConfirmation({
+          title: t`Set as global group`,
+          message: t`This group will be loaded and available in all the scenes. This is only recommended for groups that you reuse a lot and can't be undone. Make this group global?`,
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectGroupsList/index.js:298]: not sure I understand the sentence `this group will be loaded`? 
OK now I do, but it took me a while, should this be more user-friendly like `This group will appear in all scenes` or something to say it may impact performance, but not say that it's the engine's fault 😅 ?

File: newIDE/app/src/ObjectGroupsList/index.js [hunk 6]
```diff
@@ -499,4 +603,19 @@ export default class GroupsListContainer extends React.Component<Props, State> {
       </Background>
     );
   }
-}
+);
+
+const areEqual = (prevProps: Props, nextProps: Props): boolean =>
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectGroupsList/index.js:608]: Is this function rightly named? It feels like it's doing the opposite, should this be `areNotEqual` ?

Comment (ClementPasteau) [reply newIDE/app/src/ObjectGroupsList/index.js:608]: Or maybe `shouldUpdate` is better?

File: newIDE/app/src/ObjectsList/index.js [hunk 7]
```diff
@@ -48,15 +33,61 @@ import PreferencesContext from '../MainFrame/Preferences/PreferencesContext';
...
+import useAlertDialog from '../UI/Alert/useAlertDialog';
 
 const gd: libGDevelop = global.gd;
+const sceneObjectsRootFolderId = 'scene-objects';
+const globalObjectsRootFolderId = 'global-objects';
+const globalObjectsEmptyPlaceholderId = 'global-empty-placeholder';
+
+const globalObjectsWikiLink = getHelpLink(
+  '/interface/scene-editor/global-objects/',
+  ':~:text=Global%20objects%20are%20objects%20which,are%20usable%20by%20all%20Scenes'
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectsList/index.js:60]: is this link too precise? If we change the wording of the doc, this will break. Should we link to a section title instead?

File: newIDE/app/src/ObjectsList/index.js [hunk 8]
```diff
@@ -79,29 +110,79 @@ const objectTypeToDefaultName = {
...
+  if (!objectFolderOrObject.isFolder()) return null;
+  return mapFor(0, objectFolderOrObject.getChildrenCount(), i => ({
+    objectFolderOrObject: objectFolderOrObject.getChildAt(i),
+    global,
+  }));
+};
+const getTreeViewItemData = (item: TreeViewItem) =>
+  item.isRoot || item.isPlaceholder
+    ? undefined
+    : item.objectFolderOrObject.isFolder()
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectsList/index.js:150]: I believe you can move this check in the first check with a `|| item.objectFolderOrObject.isFolder()` ?

File: newIDE/app/src/ObjectsList/index.js [hunk 9]
```diff
@@ -296,17 +396,53 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
...
+            forceUpdateList();
+            selectObjectFolderOrObjectWithContext(null);
+          }
+          return;
+        }
+
+        const answer = await showDeleteConfirmation({
+          message: t`Are you sure you want to remove this object? This can't be undone.`,
+          title: t`Remove object`,
+        });
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectsList/index.js:431]: just to check, this does not require typing the name of the object to delete it right?

File: newIDE/app/src/ObjectsList/index.js [hunk 10]
```diff
@@ -418,205 +611,235 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
...
+                      label: (
+                        <Trans>
+                          There is no{' '}
+                          <Link
+                            href={globalObjectsWikiLink}
+                            onClick={() =>
+                              Window.openExternalURL(globalObjectsWikiLink)
+                            }
+                          >
+                            global object
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectsList/index.js:747]: will this get translated well? 🤔

File: newIDE/app/src/ObjectsList/index.js [hunk 11]
```diff
@@ -743,7 +1260,7 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
             label: i18n._(t`Edit effects`),
             click: () => onEditObject(object, 'effects'),
             enabled: objectMetadata.hasDefaultBehavior(
-              'EffectCapacity::EffectBehavior'
+              'EffectCapability::EffectBehavior'
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectsList/index.js:1263]: oh nice

File: newIDE/app/src/SceneEditor/index.js [hunk 12]
```diff
@@ -1136,6 +1178,73 @@ export default class SceneEditor extends React.Component<Props, State> {
...
+      label: i18n._(t`Open scene properties`),
+      click: () => this.openSceneProperties(true),
+    },
+  ];
+
+  getContextMenuInstancesWiseItems = (i18n: I18nType) => [
+    {
+      label: i18n._(t`Copy`),
+      click: () => this.copySelection(),
+      enabled: this.instancesSelection.hasSelectedInstances(),
```
Comment (ClementPasteau) [inline newIDE/app/src/SceneEditor/index.js:1196]: You can probably save `const hasSelectedInstances = this.instancesSelection.hasSelectedInstances()` at the top of this function to re-use it ad not call it for every item

File: newIDE/app/src/UI/Alert/ConfirmDeleteDialog.js [hunk 13]
```diff
@@ -75,15 +75,19 @@ function ConfirmDeleteDialog(props: Props) {
...
-          <LargeSpacer />
-          <TextField
-            autoFocus="desktop"
-            floatingLabelFixed
-            floatingLabelText={i18n._(props.fieldMessage)}
-            value={textInput}
-            onChange={(e, text) => setTextInput(text)}
-            hintText={props.confirmText}
-          />
+          {props.confirmText && props.fieldMessage && (
```
Comment (ClementPasteau) [inline newIDE/app/src/UI/Alert/ConfirmDeleteDialog.js:78]: ok I just answered my question

File: newIDE/app/src/UI/TreeView/TreeViewRow.js [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+    () => {
+      if (
+        isStayingOver &&
+        !openWhenOverTimeoutId.current &&
+        node.canHaveChildren &&
+        node.collapsed
+      ) {
+        openWhenOverTimeoutId.current = setTimeout(() => {
+          onOpen(node);
+        }, 800);
```
Comment (ClementPasteau) [inline newIDE/app/src/UI/TreeView/TreeViewRow.js:192]: should those values be extracted at the top of the file? So we can see the different time triggers for actions

File: newIDE/app/src/UI/TreeView/TreeViewRow.js [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+          connectDragSource,
+          connectDropTarget,
+          connectDragPreview,
+          isOver,
+          canDrop,
+        }) => {
+          setIsStayingOver(isOver);
+          return (
+            <div
+              style={{ paddingLeft: left }}
```
Comment (ClementPasteau) [inline newIDE/app/src/UI/TreeView/TreeViewRow.js:303]: should this be inside the `full-height-flex-container` class ?

File: newIDE/app/src/Utils/Haptic.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+// @flow
+
+import { isNativeMobileApp } from './Platform';
+
+export const hapticFeedback: ?({
+  durationInMs: number,
+}) => void = !isNativeMobileApp()
+  ? ({ durationInMs }) => {
+      if (window.navigator && window.navigator.vibrate) {
+        window.navigator.vibrate(durationInMs);
```
Comment (ClementPasteau) [inline newIDE/app/src/Utils/Haptic.js:10]: should this be try catched just for security? unsure

File: Core/GDCore/Project/Project.cpp (hunk 4 above)
Comment (AlexandreSi) [reply Core/GDCore/Project/Project.cpp:1008]: Good idea

File: Core/GDCore/Project/ObjectFolderOrObject.cpp (hunk 3 above)
Comment (AlexandreSi) [reply Core/GDCore/Project/ObjectFolderOrObject.cpp:232]: Good idea!

File: newIDE/app/src/InAppTutorial/Tutorials/OnboardingTutorial.js [hunk 17]
```diff
@@ -147,7 +147,7 @@ const inAppTutorial: InAppTutorial = {
     },
     {
       id: 'DragObjectToScene',
-      elementToHighlightId: '#object-item-0',
+      elementToHighlightId: 'objectInObjectsList:firstObject',
```
[the remark this answers is not in the document]
Comment (AlexandreSi) [reply newIDE/app/src/InAppTutorial/Tutorials/OnboardingTutorial.js:150]: You might be right yes

File: newIDE/app/src/ObjectGroupsList/index.js (hunk 5 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectGroupsList/index.js:298]: That's a good question, I did not change the copy, but it sure could be improved. @LuniMoon Do you a suggestion?

File: newIDE/app/src/ObjectsList/index.js (hunk 7 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:60]: The page is quite short so if the copy changes and the browser can't match the anchor, the impact won't be important. I would say let's keep it as is

File: newIDE/app/src/ObjectsList/index.js (hunk 9 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:431]: Indeed!

File: newIDE/app/src/ObjectsList/index.js (hunk 10 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:747]: Truth be told, I'm not sure, I'll see if we did it somewhere else (adding a link to a translated copy)

File: newIDE/app/src/SceneEditor/index.js (hunk 12 above)
Comment (AlexandreSi) [reply newIDE/app/src/SceneEditor/index.js:1196]: Thanks, I did not realize it was called for each item 😅

File: newIDE/app/src/UI/TreeView/TreeViewRow.js (hunk 14 above)
Comment (AlexandreSi) [reply newIDE/app/src/UI/TreeView/TreeViewRow.js:192]: Good idea

File: newIDE/app/src/Utils/Haptic.js (hunk 16 above)
Comment (AlexandreSi) [reply newIDE/app/src/Utils/Haptic.js]: For the sake of avoiding a crash for such a useless reason, I added it

File: newIDE/app/src/ObjectGroupsList/index.js (hunk 6 above)
Comment (ClementPasteau) [reply newIDE/app/src/ObjectGroupsList/index.js:608]: But the logic of the function actually checks when items are different, so it's doing the opposite no?

File: newIDE/app/src/ObjectGroupsList/index.js (hunk 5 above)
Comment (LuniMoon) [reply newIDE/app/src/ObjectGroupsList/index.js:298]: Suggestion:
Title: Get as global group
Message: Global elements help to manage [common? / prevalent?] objects across multiple scenes and it is recommended for the most used objects. This action cannot be undone. <break>
Do you want to set as global group? 
Set as global / Cancel

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    std::size_t newPosition) {
+  if (!newParentFolder.IsFolder()) return;
+  if (newParentFolder.IsADescendantOf(objectFolderOrObject)) return;
+
+  std::vector<std::unique_ptr<gd::ObjectFolderOrObject>>::iterator it =
+      find_if(children.begin(),
+              children.end(),
+              [&objectFolderOrObject](std::unique_ptr<gd::ObjectFolderOrObject>&
+                                          childObjectFolderOrObject) {
+                return &(*childObjectFolderOrObject) == &objectFolderOrObject;
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:174]: You can use .get() instead of deferencing and then taking the address

File: newIDE/app/src/ObjectGroupsList/index.js (hunk 6 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectGroupsList/index.js:608]: Oh yes sorry! I was looking at the ObjectsList

File: GDevelop.js/__tests__/Core.js [hunk 19]
```diff
@@ -4,6 +4,7 @@ const {
   makeFakeAbstractFileSystem,
 } = require('../TestUtils/FakeAbstractFileSystem');
 const extend = require('extend');
+const { mapFor } = require('../../newIDE/app/src/Utils/MapFor.js');
```
[the remark this answers is not in the document]
Comment (AlexandreSi) [reply GDevelop.js/__tests__/Core.js:7]: I remember doubting this move 😅

File: newIDE/app/src/EventsSheet/InstructionEditor/SelectorListItems/FolderListItem.js [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+  folderWithPath: {|
+    path: string,
+    folder: gdObjectFolderOrObject,
+    global: boolean,
+  |},
+  iconSize: number,
+  matchesCoordinates: number[][],
+|};
+
+export const renderFolderListItem = ({
```
[the remark this answers is not in the document]
Comment (AlexandreSi) [reply newIDE/app/src/EventsSheet/InstructionEditor/SelectorListItems/FolderListItem.js:17]: I'm not sure it won't have some unexpected side effects though. `renderInstructionOrExpressionListItem` is used in the `InstructionOrExpressionSelector` so I'd rather not change this

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+  return badObjectFolderOrObject;
+}
+
+void ObjectFolderOrObject::SetFolderName(const gd::String& name) {
+  if (!IsFolder()) return;
+  folderName = name;
+}
+
+ObjectFolderOrObject& ObjectFolderOrObject::GetChildAt(std::size_t index) {
+  return *children[index];
```
[the remark this answers is not in the document]
Comment (AlexandreSi) [reply Core/GDCore/Project/ObjectFolderOrObject.cpp:64]: I wondered if I should do that but I ended up mimicking what ObjectsContainer does with its objects. So I'm not sure what to do

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+ObjectFolderOrObject::~ObjectFolderOrObject() {}
+
+bool ObjectFolderOrObject::HasObjectNamed(const gd::String& name) {
+  if (IsFolder()) {
+    return std::any_of(
+        children.begin(),
+        children.end(),
+        [&name](
+            std::unique_ptr<gd::ObjectFolderOrObject>& objectFolderOrObject) {
+          return objectFolderOrObject.get()->HasObjectNamed(name);
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:37]: with a unique_ptr, you can directly use `->` like a normal pointer.

```suggestion
          return objectFolderOrObject->HasObjectNamed(name);
```

File: Core/GDCore/Project/ObjectFolderOrObject.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (4ian) [reply Core/GDCore/Project/ObjectFolderOrObject.cpp:64]: I would say: let's go for the safe option for this class and return badObjectFolderOrObject.
I prefer a logical error in the editor rather than a crash in most of these classes defining a project structure.

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+                                        std::size_t position) {
+  auto objectFolderOrObject =
+      gd::make_unique<ObjectFolderOrObject>(insertedObject, this);
+  children.insert(
+      position < children.size() ? children.begin() + position : children.end(),
+      std::move(objectFolderOrObject));
+}
+
+std::size_t ObjectFolderOrObject::GetChildPosition(
+    ObjectFolderOrObject& child) const {
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:91]:
```suggestion
    const ObjectFolderOrObject& child) const {
```

Because you don't modify the child and tell the caller "I guarantee I won't modify it, so you can pass me something that is const if you have one (and otherwise, it's the same)"

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+  for (std::size_t j = 0; j < children.size(); j++) {
+    if (children[j].get() == &child) return j;
+  }
+  return gd::String::npos;
+}
+
+ObjectFolderOrObject& ObjectFolderOrObject::InsertNewFolder(
+    const gd::String newFolderName, std::size_t position) {
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:99]:
```suggestion
    const gd::String& newFolderName, std::size_t position) {
```

(in 99% of case, you can pass a const ref to a string, even from the JS)

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+                       }),
+        children.end());
+    for (auto& it : children) {
+      it->RemoveRecursivelyObjectNamed(name);
+    }
+  }
+};
+
+bool ObjectFolderOrObject::IsADescendantOf(
+    ObjectFolderOrObject& otherObjectFolderOrObject) {
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:128]: const ref

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+  if (oldIndex >= children.size() || newIndex >= children.size()) return;
+
+  std::unique_ptr<gd::ObjectFolderOrObject> objectFolderOrObject =
+      std::move(children[oldIndex]);
+  children.erase(children.begin() + oldIndex);
+  children.insert(children.begin() + newIndex, std::move(objectFolderOrObject));
+}
+
+void ObjectFolderOrObject::RemoveFolderChild(
+    ObjectFolderOrObject& childToRemove) {
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:146]: const ref

File: Core/GDCore/Project/ObjectFolderOrObject.cpp [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+        children.push_back(std::move(childObjectFolderOrObject));
+      }
+    }
+  } else {
+    folderName = "";
+    gd::String objectName = element.GetStringAttribute("objectName");
+    if (objectsContainer.HasObjectNamed(objectName)) {
+      object = &objectsContainer.GetObject(objectName);
+    } else {
+      gd::LogError("Object with name " + objectName + " not found in objects container.");
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.cpp:236]: Also set object to nullptr for safety

File: Core/GDCore/Project/ObjectFolderOrObject.h [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+  ///@}
+
+ private:
+  static gd::ObjectFolderOrObject badObjectFolderOrObject;
+
+  gd::ObjectFolderOrObject*
+      parent;  // nullptr if root folder, points to the parent folder otherwise.
+
+  // Representing an object:
+  gd::Object* object;  // Empty if folderName is set.
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.h:187]: nit:
```suggestion
  gd::Object* object;  // nullptr if folderName is set.
```

File: Core/GDCore/Project/ObjectFolderOrObject.h [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+  /**
+   * \brief Inserts an instance representing the given object at the end of the
+   * children.
+   */
+  void InsertObject(gd::Object* insertedObject);
+  /**
+   * \brief Inserts an instance representing the given object at the given
+   * position.
+   */
+  void InsertObject(gd::Object* insertedObject, std::size_t position);
```
Comment (4ian) [inline Core/GDCore/Project/ObjectFolderOrObject.h:136]:
```suggestion
  /**
   * \brief Inserts an instance representing the given object at the given
   * position.
   */
  void InsertObject(gd::Object* insertedObject, std::size_t position = (size_t)-1);
```

which means that if you don't pass position, C++ will replace it with the maximum position (because size_t is unsigned, so -1 is the max value), which will insert it at the end (you have to adapt the method to handle this)

File: newIDE/app/src/EventsSheet/InstructionEditor/SelectorListItems/FolderListItem.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/SelectorListItems/FolderListItem.js:17]: Let's not change it for now then

File: newIDE/app/src/ObjectsList/index.js [hunk 30]
```diff
@@ -418,205 +609,235 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
...
           objectName: name,
-          positionObjectName: name,
+          positionObjectFolderOrObjectWithContext: objectFolderOrObjectWithContext,
           objectType: type,
           serializedObject,
-          global,
         });
 
-        editName(newObjectWithContext);
+        const newObjectFolderOrObjectWithContext = {
```
Comment (4ian) [inline newIDE/app/src/ObjectsList/index.js:660]: Question: from your experience working on this @AlexandreSi, do you think we're far from being able to remove entirely this "WithContext" intermediate object. I.e: where is `global` used and what do we need to entirely remove our reliance on it?

(this is not something to address in this PR)

Comment (4ian) [reply newIDE/app/src/ObjectsList/index.js:660]: (it's possible that we can't remove it, but one day we will rework `global` so that instead of being true/false, it will be a reference to the ObjectsContainer - so we can work with any number of "scope" of objects, in case we have folder of scenes)

File: newIDE/app/src/ObjectsList/index.js (hunk 10 above)
Comment (4ian) [reply newIDE/app/src/ObjectsList/index.js:745]: You could consider using markdown + a placeholder for the link `There is no [globla object](LINK_PLACEHOLDER_DO_NOT_TRANSLATE_THIS) yet.`

I think we did something like this, but it's a bit hacky/fragile.

File: newIDE/app/src/ObjectsList/index.js [hunk 31]
```diff
@@ -637,94 +858,390 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
...
+                      0
+                    );
+                  onObjectModified(true);
+                  if (treeViewRef.current)
+                    treeViewRef.current.openItems([
+                      getTreeViewItemId({
+                        objectFolderOrObject: folder,
+                        global: item.global,
+                      }),
+                    ]);
```
Comment (4ian) [inline newIDE/app/src/ObjectsList/index.js:1176]: Could almost be extracted as a function, it's a bit of logic buried inside a menu.

File: newIDE/app/src/ObjectsList/index.js [hunk 32]
```diff
@@ -809,90 +1331,147 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
...
+              objectFolderOrObject: item.objectFolderOrObject.getChildAt(0),
+              global: item.global,
+            };
+          }
+        },
+        onArrowLeft: (item, isOpened) => {
+          if (item.isPlaceholder || item.isRoot) return null;
+          if (!item.objectFolderOrObject.isFolder() || !isOpened) {
+            const parent = item.objectFolderOrObject.getParent();
+            if (parent.getFolderName() === '__ROOT') return null;
```
Comment (4ian) [inline newIDE/app/src/ObjectsList/index.js:1390]: Would be maybe worth using a method `parent.IsRootFolder()`. The C++ class can tell you, it just has to check if it's a folder that has no parent.

The risk of using __ROOT is that this will break when I will use "__ROOT" as the name, and this is not type safe (this will break when I rework the code and type `_ROOT` instead of `__ROOT`).

Comment (4ian) [reply newIDE/app/src/ObjectsList/index.js:1390]: (this is a way to say "no magic strings" :))

File: newIDE/app/src/SceneEditor/index.js [hunk 33]
```diff
@@ -1261,27 +1335,24 @@ export default class SceneEditor extends React.Component<Props, State> {
           label: i18n._(t`Edit behaviors`),
           click: () => this.editObjectByName(objectName, 'behaviors'),
         },
-        {
-          label: i18n._(t`Edit effects`),
-          click: () => this.editObjectByName(objectName, 'effects'),
-        },
-      ];
+        objectMetadata
```
Comment (4ian) [inline newIDE/app/src/SceneEditor/index.js:1338]: Nice

File: newIDE/app/src/UI/DragAndDrop/DragAndDropContextProvider.js (hunk 2 above)
Comment (4ian) [reply newIDE/app/src/UI/DragAndDrop/DragAndDropContextProvider.js:7]: Coming back to this: would be worth a comment I think

File: newIDE/app/src/UI/TreeView/index.js [hunk 34]
```diff
@@ -0,0 +1,609 @@
...
+            delta -= 1;
+            newFocusedNode = flattenedData[itemIndexInFlattenedData + delta];
+          }
+          if (newFocusedNode) {
+            newFocusedItem = newFocusedNode.item;
+          }
+        }
+      } else if (event.key === 'ArrowRight' && arrowKeyNavigationProps) {
+        event.preventDefault();
+        newFocusedItem = arrowKeyNavigationProps.onArrowRight(
```
Comment (4ian) [inline newIDE/app/src/UI/TreeView/index.js:547]: I wonder if we could give a tiny bit less responsibility to the parent:

onArrowRight can either:
- return another item to focus (which is what it does now)
- or return null. In this case, we `closeItems`/`openItems` according to if it's already opened or not. 

I just find it a bit weird that the parent ObjectsList does `treeViewRef.current.openItems([getTreeViewItemId(item)]);` while it could be handled by the tree... the tree needs just to know if there is a children or parent to navigate to, otherwise it can handle closing/opening

File: newIDE/app/src/ObjectsList/index.js (hunk 30 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:660]: Well, `global` is only used to know which container operations should be done on. I haven't really thought about the idea of removing this but it could be more convenient indeed

File: newIDE/app/src/ObjectsList/index.js (hunk 10 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:745]: We already used the syntax `<Trans>...<Link>...</Link>...</Trans>` here: https://github.com/4ian/GDevelop/blob/master/newIDE/app/src/MainFrame/Preferences/PreferencesDialog.js#L151-L166

And it seems to work fine!

File: GDevelop.js/__tests__/Core.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (ClementPasteau) [reply GDevelop.js/__tests__/Core.js:7]: I'd recommend duplicating it too?

File: newIDE/app/src/ObjectGroupsList/index.js [hunk 35]
```diff
@@ -67,430 +80,524 @@ type Props = {|
...
+          },
+          {
+            label: i18n._(t`Scene Groups`),
+            children:
+              objectGroupsList.length > 0
+                ? objectGroupsList
+                : [
+                    {
+                      label: i18n._(t`Start by adding a new group.`),
+                      id: 'scene-empty-placeholder',
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectGroupsList/index.js:506]: why is this ID not in a variable like the others?

File: newIDE/app/src/ObjectGroupsList/index.js [hunk 36]
```diff
@@ -67,430 +80,524 @@ type Props = {|
...
+                          selectedGroupWithContext
+                            ? [selectedGroupWithContext]
+                            : []
+                        }
+                        onSelectItems={items => {
+                          if (!items) setSelectedGroupWithContext(null);
+                          const itemToSelect = items[0];
+                          if (
+                            'isRoot' in itemToSelect ||
+                            'isPlaceholder' in itemToSelect
```
Comment (ClementPasteau) [inline newIDE/app/src/ObjectGroupsList/index.js:576]: those strings are a bit magic, should they be in a constant somewhere? (you're using it in multiple places too)

File: newIDE/app/src/UI/TreeView/TreeViewRow.js [hunk 37]
```diff
@@ -0,0 +1,461 @@
...
+        item: node.item,
+        x: clientX,
+        y: clientY,
+      });
+    },
+    [onContextMenu, index, node.item]
+  );
+
+  const longTouchForContextMenuProps = useLongTouch(openContextMenu, {
+    delay: 1000,
```
Comment (ClementPasteau) [inline newIDE/app/src/UI/TreeView/TreeViewRow.js:158]: constant this number too?

File: newIDE/app/src/UI/TreeView/TreeViewRow.js (hunk 15 above)
Comment (ClementPasteau) [reply newIDE/app/src/UI/TreeView/TreeViewRow.js:306]: Oh I completely missed that it was not a fixed value.

File: newIDE/app/src/ObjectGroupsList/index.js (hunk 35 above)
Comment (AlexandreSi) [reply newIDE/app/src/ObjectGroupsList/index.js:506]: Because I didn't need to use it anywhere else, but I will variabalize it

**Review by 4ian — APPROVED** [review]

File: newIDE/app/src/UI/TreeView/index.js [hunk 38]
```diff
@@ -401,6 +408,22 @@ const TreeView = <Item: ItemBaseAttributes>(
...
+    (item: Item) => {
+      setAnimatedItemId(getItemId(item));
+    },
+    [getItemId]
+  );
+
+  React.useEffect(
+    () => {
+      if (animatedItemId) {
+        setTimeout(() => setAnimatedItemId(''), 400);
```
Comment (4ian) [inline newIDE/app/src/UI/TreeView/index.js]: In theory you should cancel the timeout if the animatedItemId is changed, so that one can:

- Animate item 1
- 200ms after, Animate item 2 (item 1 stops animating)
- 200ms after, the animated item id is NOT reset (because it's the timeout of the first animatedItemId)
- 200ms after, the animated item id IS reset (it's the second timeout, and it's perfectly fine and does not need to be cancelled).

**Review by 4ian — APPROVED** [review]

File: newIDE/app/src/UI/TreeView/index.js (hunk 38 above)
Comment (AlexandreSi) [reply newIDE/app/src/UI/TreeView/index.js]: I was lazy and thought that 400ms were too quick for anything like this to happen 😅 
It's now fixed
